File: js/countConstruct_tab.js
// countConstruct(target, wordBank)
//
// countConstruct(purple, [purp, p, ur, le, purpl]) -> 2


function countConstruct(target, wordBank) {
	let table = Array(target.length + 1).fill(0)
	table[0] = 1

	for (let i=0; i<=target.length; i++) {
		if (target[i] !== 0)
			for (let word of wordBank) {
				const size = word.length
				if (target.slice(i, size + i) === word) {
					table[size + i] += table[i]
				}
			}
	}

	// console.log(table)
	return table[target.length]
}


console.log(countConstruct('purple', ['purp', 'p', 'ur', 'le', 'purpl']))

File: cpp/countConstruct_tab.cpp
#include <iostream>
#include <vector>

using namespace std;

int countConstruct(string target, vector<string>& wordBank) {
	vector<int> table (target.size() + 1, 0);
	table[0] = 1;

	for (int i=0; i<=target.size(); i++) {
		if (table[i] != 0)
			for (auto word : wordBank) {
				if (target.compare(i, word.size() + i, word) == 0) {
					table[word.size() + i] += table[i];
				}
			}
	}

	return table[target.size()];
}

int countConstruct(string target, vector<string>&& wordBank) {
	return countConstruct(target, wordBank);
}

int main() {
	cout << countConstruct("abcdef", {"ef", "ab", "bc", "abc", "def", "abcd"}) << endl;
	cout << countConstruct("skateboard", {"ab", "bc", "abc", "def", "abcd"}) << endl;
}
